rfcs/0132-meson-builds-nix.md

```diff
@@ -0,0 +1,216 @@
+
+## Learning the new system
+
+A somewhat subjective but non-negligible issue is the entry barrier of this new build system. Switching from a known build system to one unknown is not without its problems.
+
+However, the Meson development team strives to keep the DSL easy to learn and pleasurable to use. It should not be hard to become familiar with the Python-esque syntax of meson, and its functional, stateless approach is certainly a feature highly appreciated by the Nix community as a whole.
+
+The huge advantages of implementing Meson surpass the small drawbacks of learning it.
+
+## Source code changes
+
+Further to the obvious inclusion of meson files (and the removal of the old quasi-autotools ones), there is a reasonable expectation of code refactoring.
+
+However, such refactorings are completely validated on the long term goals of Nix, in particular the improvements on portability.
+
+## End users
+
+The most known end user of Nix is certainly Nixpkgs. However, there are many other Linux distributions that already keep Nix on their repositories (15 families, according to Repology). There is also a reasonable expectation of affecting those package managers' devteams.
+
+However, most (if not all) of those distributions already have Meson and its companion tool Ninja in their respective package databases (53 families, according to Repology), given that many open source projects use them as build system.
+
+## Transition between old and new build infrastructure
+
+The transition between between the old and new build systems should be as smooth and controlled as possible.
+
+# Alternatives
+[alternatives]: #alternatives
+
+The alternatives are
+
+- Do nothing
+
+  It would keep the current code confusing and harder to work with, as stated on the motivation section.
+
+- Use CMake
+
+  Indeed, CMake has many noteworthy advantages:
+
+  - Supports typical high level idiomatic constructions.
+  - Can generate GMake- and Ninja-compatible scripts.
+    - By design, Meson does not provide a Make backend.
+  - Both Meson and CMake support Microsoft Windows NT and Apple MacOS platforms
+    - As well as MSVC and XCode programming environments.
+
+  - On the other hand, CMake DSL cmake language is arguably more complex and cumbersome.
+
+- Evaluate other building systems (waf, premake, bazel, xmake etc.)
+
+  About this, a principle should be observed:
+
+  Per the bootstrap route issue discussed above, build tools strongly tied to other programming languages are strongly discouraged.
+
+  E.g. waf is basically a Python library, whereas premake and xmake are Lua libraries. They can't be decoupled of their implementation languages.
+
+- Use Bazel
+
+  For the sake of completeness, there is Bazel.
+
+  Advantages:
+
+  - Fast, scalable, multi-language and extensible.
```

Comment:
I object to this statement on the grounds that in my experience, running Bazel was significantly slower than Meson or cmake or autotools.

I suspect that when Bazel advertises itself as "fast", it actually means "well optimized for the use case of millions of translation units", in other words the "fast" is a qualifier for "scalable" and basically means that if you are Google and have to build the entirety of Google's entire international network across billions of machines, Bazel is still competitively fast.

This is, of course, the most important criterion for an open source project with 189 translation units.

Incidentally, the most compelling argument against Bazel, which is that "Google's problems are not your problems, Googleware is usually not your best option, and doing something because Google did it won't make you successful like Google", is just...

... a generic version of that argument.

Reply:
They themselves say they are "Fast, scalable, multi-language and extensible" on their own homepage. I merely copy-pasted them.

This section was written for the sake of completeness.

Reply:
:p


Of course Google is tooting their own horn. Like I said, I don't buy into their hype; I think they have very different definitions of "fast" than most people.

Reply:
Reworded.